File: misc/codechef/june_3.cpp
#include <iostream>
#include <stdio.h>
#include <cmath>
#include <algorithm>
#include <cstring>
#include <string>
#include <ctime>
#include <vector>
#define gc getchar_unlocked

using namespace std;

inline int fast_io() 
{
	register int N = 0, C;
	while ((C = gc()) < '0');
 
	do {
		N = (N<<3) + (N<<1) + C - '0';
	}while ((C = gc()) >= '0');
	return N;
}


int main(){
	int test;
	test = fast_io();
	
	long int a[10000];
	long int b[10000];
	long int max1[10000];
	long int max2[10000];
	long int min1[10000];
	long int min2[10000];

	while(test--){
		long int n;
		n = fast_io();

		for(int i=0;i<n;i++){
			scanf("%ld",&a[i]);
			b[i]=-a[i];
		}

		long int maxendhere=a[0];
		long int maxsofar=a[0];
		max1[0]=a[0];

		for(int i=1;i<n;i++){
			maxendhere=maxendhere+a[i];
			maxendhere = maxendhere > a[i] ? maxendhere:a[i];
			maxsofar = maxsofar > maxendhere ? maxsofar : maxendhere;
			max1[i]=maxsofar;
		}

/*		for(int i=0;i<n;i++)
			cout<<max1[i]<<" ";
			cout<<endl;
*/
		maxendhere = max2[n-1] = maxsofar = a[n-1];
//		cout<<max2[n-1]<<endl;
		for(int i=n-2;i>=0;i--){
				maxendhere=maxendhere+a[i];
				maxendhere = maxendhere > a[i] ? maxendhere:a[i];
				maxsofar = maxsofar > maxendhere ? maxsofar : maxendhere;
				max2[i]=maxsofar;	
		}

		maxendhere = maxsofar = b[0];

		min1[0]=-b[0];

		for(int i=1;i<n;i++){
			maxendhere=maxendhere+b[i];
			maxendhere = maxendhere > b[i] ? maxendhere:b[i];
			maxsofar = maxsofar > maxendhere ? maxsofar : maxendhere;
			min1[i]= -maxsofar;
		}


		maxendhere = maxsofar = b[n-1];

		min2[n-1] = -b[n-1];

		for(int i=n-2;i>=0;i--){
			maxendhere=maxendhere+b[i];
			maxendhere = maxendhere > b[i] ? maxendhere:b[i];
			maxsofar = maxsofar > maxendhere ? maxsofar : maxendhere;
			min2[i]= -maxsofar;	
		}

/*		for(int i=0;i<n;i++){
			cout<<a[i]<<" "<<b[i]<<" "<<max1[i]<<" "<<max2[i]<<" "<<min1[i]<<" "<<min2[i]<<endl;
		}
*/

		long int main_max = 0;

		for(int i=1;i<n;i++){
			long m1 = abs(max1[i-1]-min2[i]);
			long m2 = abs(max2[i]-min1[i-1]);

			m1 = m1 > m2 ? m1:m2;
			main_max = main_max > m1 ? main_max : m1;
		}
		printf("%ld\n",main_max);
	}	
	return 0;
}

File: misc/codechef/june_3.java
import java.io.*;
public class Main {
	
	public static void main(String args[]) throws IOException
	{
		int T;
		BufferedReader br=new BufferedReader(new InputStreamReader(System.in));
		T=Integer.parseInt(br.readLine());
		long a[]=new long[10000];
		long b[]=new long[10000];
		long max1[]=new long[10000];
		long min1[]=new long[10000];
		long max2[]=new long[10000];
		long min2[]=new long[10000];
		while(T-->0)
		{
			int n=Integer.parseInt(br.readLine());
			String arr[]=br.readLine().split(" " );
			int i=0;
			for(String s:arr)
			{
				a[i]=Integer.parseInt(arr[i]);
				b[i]=-a[i];
				i++;
			}
			
			long maxsofar=a[0],maxendinghere=a[0];max1[0]=a[0];
			for(i=1;i<n;i++)
			{
					//maxendinghere=maxendinghere+a[i]>0?maxendinghere+a[i]:0;
					maxendinghere=maxendinghere+a[i];
					maxendinghere=maxendinghere>a[i]?maxendinghere:a[i];
					maxsofar=maxsofar>maxendinghere?maxsofar:maxendinghere;
					max1[i]=maxsofar;
			}
		
			max2[n-1]=maxendinghere=maxsofar=a[n-1];
			for(i=n-2;i>=0;i--)
			{
					//maxendinghere=maxendinghere+a[i]>0?maxendinghere+a[i]:0;
					maxendinghere=maxendinghere+a[i];
					maxendinghere=maxendinghere>a[i]?maxendinghere:a[i];
					maxsofar=maxsofar>maxendinghere?maxsofar:maxendinghere;
					max2[i]=maxsofar;
				
			}
			maxendinghere=maxsofar=b[0];
			min1[0]=-b[0];
			
			for(i=1;i<n;i++)
			{
					//maxendinghere=maxendinghere+b[i]>0?maxendinghere+b[i]:0;
					maxendinghere=maxendinghere+b[i];
					maxendinghere=maxendinghere>b[i]?maxendinghere:b[i];
					maxsofar=maxsofar>maxendinghere?maxsofar:maxendinghere;
					min1[i]=-maxsofar;
			}


		/*	for(i=0;i<n;i++)
				System.out.print(min1[i]+" ");
				System.out.println();
		*/
			maxendinghere=maxsofar=b[n-1];
			min2[n-1]=-b[n-1];
			
			for(i=n-2;i>=0;i--)
			{
					//maxendinghere=maxendinghere+b[i]>0?maxendinghere+b[i]:0;
					maxendinghere=maxendinghere+b[i];
					maxendinghere=maxendinghere>b[i]?maxendinghere:b[i];
					maxsofar=maxsofar>maxendinghere?maxsofar:maxendinghere;
					min2[i]=-maxsofar;
			}

		/*	for(i=0;i<n;i++)
				System.out.print(min2[i]+" ");
				System.out.println();
		*/	
			long globalmax=0;
			
		/*	for(i=0;i<n;i++)
			{
				System.out.println(a[i]+" "+b[i]+" "+min1[i]+" "+max1[i]+" "+min2[i]+" "+max2[i]);
			}
		*/	
			for(i=1;i<n;i++)
			{
				long m1=max1[i-1]-min2[i];
				m1=Math.abs(m1);
				long m2=max2[i]-min1[i-1];
				m2=Math.abs(m2);
				m1=m1>m2?m1:m2;
				globalmax=globalmax>m1?globalmax:m1;
			}
			System.out.println(globalmax);
		}
		
	}

}